stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java

```diff
@@ -143,107 +156,276 @@ public BStruct generateNext() {
+                                bStruct.setFloatField(++doubleRegIndex, dValue);
+                            } else {
+                                handleNilToNonNilableFieldAssignment();
+                            }
+                        }
                         break;
                     case Types.NUMERIC:
                     case Types.DECIMAL:
                         double decimalValue = 0;
                         BigDecimal bigDecimalValue = rs.getBigDecimal(index);
+                        isOriginalValueNull = rs.wasNull();
                         if (bigDecimalValue != null) {
                             decimalValue = bigDecimalValue.doubleValue();
                         }
-                        bStruct.setFloatField(++doubleRegIndex, decimalValue);
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            BRefType refValue = isOriginalValueNull ? null : new BFloat(decimalValue);
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex, refValue,
+                                    TypeTags.FLOAT_TAG);
+                        } else {
+                            if (!isOriginalValueNull) {
+                                bStruct.setFloatField(++doubleRegIndex, decimalValue);
+                            } else {
+                                handleNilToNonNilableFieldAssignment();
+                            }
+                        }
                         break;
                     case Types.BIT:
                     case Types.BOOLEAN:
                         boolean boolValue = rs.getBoolean(index);
-                        bStruct.setBooleanField(++booleanRegIndex, boolValue ? 1 : 0);
+                        isOriginalValueNull = rs.wasNull();
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            BRefType refValue = isOriginalValueNull ? null : new BBoolean(boolValue);
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex, refValue,
+                                    TypeTags.BOOLEAN_TAG);
+                        } else {
+                            if (!isOriginalValueNull) {
+                                bStruct.setBooleanField(++booleanRegIndex, boolValue ? 1 : 0);
+                            } else {
+                                handleNilToNonNilableFieldAssignment();
+                            }
+                        }
                         break;
                     case Types.STRUCT:
                         Struct structdata = (Struct) rs.getObject(index);
                         BType structFieldType = this.type.getStructFields()[index - 1].getFieldType();
-                        fieldType = structFieldType.getTag();
-                        if (fieldType == TypeTags.STRUCT_TAG) {
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
```

Comment:
Shall we change this to BType structFieldType = structFields[index - 1].getFieldType();